mshadow/spatial_max_unpool_forward.h

```diff
@@ -0,0 +1,137 @@
+      : mask_(mask), src_(src), ksize_y_(ksize_y), ksize_x_(ksize_x), kstride_(kstride) {
+    this->src_height_ = src_shape[0];
+    this->src_width_  = src_shape[1];
+    this->out_height_ = out_shape[0];
+    this->out_width_ = out_shape[1];
+  }
+};
+
+template<typename IndexExp, typename SrcExp,
+         typename DType, int e1, int e2>
+inline MaxUnpoolingForwardExp<IndexExp, SrcExp, default_real_t>
+max_unpool_forward(const Exp<IndexExp, DType, e1> &mask, const Exp<SrcExp, DType, e2> &src, 
+  Shape<2> src_shape, Shape<2> out_shape,
+  index_t ksize_y, index_t ksize_x, index_t kstride) {
+  TypeCheckPass<ExpInfo<MaxUnpoolingForwardExp<IndexExp, SrcExp, DType> >::kDim >= 2>
+      ::Error_Expression_Does_Not_Meet_Dimension_Req();
+  return MaxUnpoolingForwardExp<IndexExp, SrcExp, default_real_t>
+      (mask.self(), src.self(), src_shape, out_shape, ksize_y, ksize_x, kstride);
+}
+// Execution plan
+template<typename IndexExp, typename SrcExp, typename DType>
+struct Plan<MaxUnpoolingForwardExp<IndexExp, SrcExp, DType>, DType> {
+ public:
+  explicit Plan(const MaxUnpoolingForwardExp<IndexExp, SrcExp, DType> &e)
+      : mask_(MakePlan(e.mask_)), src_(MakePlan(e.src_)),
+        ksize_y_(e.ksize_y_), ksize_x_(e.ksize_x_), kstride_(e.kstride_),
+        src_height_(e.src_height_), src_width_(e.src_width_),
+        out_height_(e.out_height_), out_width_(e.out_width_) {}
+  MSHADOW_XINLINE DType Eval(index_t i, index_t j) const {
+    using namespace std;
+    using namespace red;
+    const index_t y = i % out_height_;
+    const index_t c = i / out_height_;
+    const index_t x = j;
+    const index_t py_min =
+        y < ksize_y_ ? 0 : (y - ksize_y_ + kstride_) / kstride_;
+    const index_t px_min =
+        x < ksize_x_ ? 0 : (x - ksize_x_ + kstride_) / kstride_;
+    const index_t py_max = min((y + kstride_) / kstride_, src_height_);
+    const index_t px_max = min((x + kstride_) / kstride_, src_width_);
+    const index_t current_i = (c * out_height_ + y) * out_width_ + x;
+    
+    index_t index = 0;
+    for (index_t py = py_min; py < py_max; ++py) {
+      for (index_t px = px_min; px < px_max; ++px) {
+        index = static_cast<index_t>(mask_.Eval(c * src_height_ + py, px));
+        if(current_i == index) {
+          return src_.Eval(c * src_height_ + py, px);
+        }
+      }
+    }
+    return (DType)(0);
+    // return limits::MinValue<DType>();
+  }
+ private:
+  Plan<IndexExp, DType> mask_;
+  Plan<SrcExp, DType> src_;
+  const index_t ksize_y_, ksize_x_, kstride_;
+  const index_t src_height_, src_width_;
+  const index_t out_height_, out_width_;
+};
+
+template<typename IndexExp, typename SrcExp, typename DType>
+inline Plan<MaxUnpoolingForwardExp<IndexExp, SrcExp, DType>, DType>
+MakePlan(const MaxUnpoolingForwardExp<IndexExp, SrcExp, DType> &exp) {
+  return Plan<MaxUnpoolingForwardExp<IndexExp, SrcExp, DType>, DType>(exp);
+}
+
+// TODO: use shapecheck to pass shape parameter
+// TODO: remove the magic number 4
```

Comment:
1. for the fitst line:
Because currently I'm directly passing the shapes of  the data from the layer to mshadow. But in /extension/spatial_upsampling_nearest.h I found they are using ShapeCheck() to pass some of the shapes. I was thinking about imitating their way. This isn't an important TODO, just a TODO for improving,
2. as for the second line:
Because all the data used in network are 4-dimension(n,c,h,w). I wasn't sure if this layer could be applied to other forms of data other than the 4-dimension. But currently I will remove this TODO since 4-dimension data are all we need.